File: python/backtest.py
"""
Backtesting Framework with Attribution Logging

This module provides backtesting capabilities for trading models
with integrated gradients attribution tracking.
"""

import numpy as np
import pandas as pd
from typing import Optional, List, Dict, Tuple, Union
from dataclasses import dataclass, field
from datetime import datetime
import warnings


@dataclass
class Trade:
    """Represents a single trade."""
    entry_time: datetime
    exit_time: Optional[datetime]
    direction: int  # 1 for long, -1 for short
    entry_price: float
    exit_price: Optional[float]
    size: float
    pnl: Optional[float]
    attributions: Optional[np.ndarray]
    prediction_confidence: float


@dataclass
class BacktestResults:
    """Container for backtest results."""
    total_return: float
    sharpe_ratio: float
    sortino_ratio: float
    max_drawdown: float
    win_rate: float
    profit_factor: float
    total_trades: int
    avg_trade_pnl: float
    equity_curve: pd.Series
    trades: List[Trade]
    attribution_analysis: Dict


class IGBacktester:
    """
    Backtester with integrated gradients attribution logging.

    Tracks feature attributions for each trade to analyze
    which features drive profitable vs. losing trades.

    Parameters
    ----------
    model : TradingModelWithIG
        Trading model with IG capability
    initial_capital : float
        Starting capital
    transaction_cost : float
        Transaction cost as fraction of trade value
    position_size : float
        Position size as fraction of capital
    log_attributions : bool
        Whether to log attributions for each trade
    attribution_threshold : float
        Only log features with |attribution| > threshold
    """

    def __init__(
        self,
        model,
        initial_capital: float = 100_000,
        transaction_cost: float = 0.001,
        position_size: float = 0.1,
        log_attributions: bool = True,
        attribution_threshold: float = 0.05,
    ):
        self.model = model
        self.initial_capital = initial_capital
        self.transaction_cost = transaction_cost
        self.position_size = position_size
        self.log_attributions = log_attributions
        self.attribution_threshold = attribution_threshold

        self.trades: List[Trade] = []
        self.equity_curve: List[float] = []
        self.attribution_log: List[Dict] = []

    def _generate_signal(
        self,
        features: np.ndarray,
        threshold: float = 0.5,
    ) -> Tuple[int, float, Optional[np.ndarray]]:
        """
        Generate trading signal from model prediction.

        Returns
        -------
        signal : int
            Trading signal: 1 (long), -1 (short), 0 (neutral)
        confidence : float
            Prediction confidence
        attributions : np.ndarray or None
            Feature attributions if logging enabled
        """
        if self.log_attributions:
            pred, attr = self.model.predict_with_explanations(
                features.reshape(1, -1),
                return_proba=True
            )
            attributions = attr[0]
        else:
            pred = self.model.predict(features.reshape(1, -1), return_proba=True)
            attributions = None

        confidence = pred[0, 0] if pred.ndim > 1 else pred[0]

        if confidence > threshold:
            signal = 1  # Long
        elif confidence < (1 - threshold):
            signal = -1  # Short
        else:
            signal = 0  # Neutral

        return signal, confidence, attributions

    def run(
        self,
        data: pd.DataFrame,
        features_cols: List[str],
        price_col: str = "close",
        timestamp_col: str = "timestamp",
        signal_threshold: float = 0.55,
        holding_period: int = 1,
    ) -> BacktestResults:
        """
        Run backtest on historical data.

        Parameters
        ----------
        data : pd.DataFrame
            Historical data with features and prices
        features_cols : list of str
            Column names for features
        price_col : str
            Column name for price
        timestamp_col : str
            Column name for timestamp
        signal_threshold : float
            Threshold for generating trading signals
        holding_period : int
            Number of bars to hold position

        Returns
        -------
        results : BacktestResults
            Backtest results
        """
        self.trades = []
        self.equity_curve = [self.initial_capital]
        self.attribution_log = []

        capital = self.initial_capital
        position = None
        bars_held = 0

        for i in range(len(data) - holding_period):
            row = data.iloc[i]
            features = row[features_cols].values.astype(np.float32)
            current_price = row[price_col]
            current_time = row[timestamp_col] if timestamp_col in row else i

            # Check if we need to close position
            if position is not None:
                bars_held += 1
                if bars_held >= holding_period:
                    # Close position
                    exit_price = current_price
                    pnl = position.direction * (exit_price - position.entry_price) / position.entry_price
                    pnl -= self.transaction_cost  # Exit cost

                    position.exit_time = current_time
                    position.exit_price = exit_price
                    position.pnl = pnl * position.size * capital

                    capital += position.pnl
                    self.trades.append(position)
                    position = None
                    bars_held = 0

            # Generate signal if no position
            if position is None:
                signal, confidence, attributions = self._generate_signal(
                    features, signal_threshold
                )

                if signal != 0:
                    # Open position
                    trade_capital = capital * self.position_size
                    entry_cost = trade_capital * self.transaction_cost

                    position = Trade(
                        entry_time=current_time,
                        exit_time=None,
                        direction=signal,
                        entry_price=current_price,
                        exit_price=None,
                        size=self.position_size,
                        pnl=None,
                        attributions=attributions,
                        prediction_confidence=confidence,
                    )

                    capital -= entry_cost

                    if self.log_attributions and attributions is not None:
                        self.attribution_log.append({
                            'time': current_time,
                            'direction': signal,
                            'confidence': confidence,
                            'attributions': attributions,
                        })

            self.equity_curve.append(capital)

        # Close any remaining position
        if position is not None:
            exit_price = data.iloc[-1][price_col]
            pnl = position.direction * (exit_price - position.entry_price) / position.entry_price
            pnl -= self.transaction_cost

            position.exit_time = data.iloc[-1][timestamp_col] if timestamp_col in data.columns else len(data) - 1
            position.exit_price = exit_price
            position.pnl = pnl * position.size * capital

            capital += position.pnl
            self.trades.append(position)

        # Calculate metrics
        results = self._calculate_metrics()
        return results

    def _calculate_metrics(self) -> BacktestResults:
        """Calculate backtest performance metrics."""
        equity = pd.Series(self.equity_curve)
        returns = equity.pct_change().dropna()

        # Total return
        total_return = (equity.iloc[-1] / equity.iloc[0]) - 1

        # Sharpe ratio (assuming 252 trading days)
        if returns.std() > 0:
            sharpe_ratio = returns.mean() / returns.std() * np.sqrt(252)
        else:
            sharpe_ratio = 0.0

        # Sortino ratio
        downside_returns = returns[returns < 0]
        if len(downside_returns) > 0 and downside_returns.std() > 0:
            sortino_ratio = returns.mean() / downside_returns.std() * np.sqrt(252)
        else:
            sortino_ratio = sharpe_ratio

        # Maximum drawdown
        peak = equity.cummax()
        drawdown = (equity - peak) / peak
        max_drawdown = drawdown.min()

        # Trade statistics
        if self.trades:
            pnls = [t.pnl for t in self.trades if t.pnl is not None]
            wins = [p for p in pnls if p > 0]
            losses = [p for p in pnls if p <= 0]

            win_rate = len(wins) / len(pnls) if pnls else 0
            profit_factor = sum(wins) / abs(sum(losses)) if losses and sum(losses) != 0 else float('inf')
            avg_trade_pnl = np.mean(pnls) if pnls else 0
        else:
            win_rate = 0
            profit_factor = 0
            avg_trade_pnl = 0

        # Attribution analysis
        attribution_analysis = self._analyze_attributions()

        return BacktestResults(
            total_return=total_return,
            sharpe_ratio=sharpe_ratio,
            sortino_ratio=sortino_ratio,
            max_drawdown=max_drawdown,
            win_rate=win_rate,
            profit_factor=profit_factor,
            total_trades=len(self.trades),
            avg_trade_pnl=avg_trade_pnl,
            equity_curve=equity,
            trades=self.trades,
            attribution_analysis=attribution_analysis,
        )

    def _analyze_attributions(self) -> Dict:
        """Analyze attributions for winning vs. losing trades."""
        if not self.trades or not self.log_attributions:
            return {}

        # Separate winning and losing trades
        winning_attrs = []
        losing_attrs = []

        for trade in self.trades:
            if trade.attributions is None or trade.pnl is None:
                continue

            if trade.pnl > 0:
                winning_attrs.append(trade.attributions)
            else:
                losing_attrs.append(trade.attributions)

        analysis = {
            'n_winning_trades': len(winning_attrs),
            'n_losing_trades': len(losing_attrs),
        }

        if winning_attrs:
            winning_attrs = np.array(winning_attrs)
            analysis['winning_mean_attr'] = winning_attrs.mean(axis=0)
            analysis['winning_std_attr'] = winning_attrs.std(axis=0)

        if losing_attrs:
            losing_attrs = np.array(losing_attrs)
            analysis['losing_mean_attr'] = losing_attrs.mean(axis=0)
            analysis['losing_std_attr'] = losing_attrs.std(axis=0)

        if winning_attrs and losing_attrs:
            # Discriminative features (difference between winners and losers)
            analysis['discriminative_attr'] = (
                analysis['winning_mean_attr'] - analysis['losing_mean_attr']
            )

        return analysis

    def get_top_features(
        self,
        n_features: int = 10,
        feature_names: Optional[List[str]] = None,
    ) -> pd.DataFrame:
        """
        Get top features by attribution for winning trades.

        Parameters
        ----------
        n_features : int
            Number of top features to return
        feature_names : list of str, optional
            Feature names

        Returns
        -------
        df : pd.DataFrame
            Top features with attributions
        """
        analysis = self._analyze_attributions()

        if 'winning_mean_attr' not in analysis:
            return pd.DataFrame()

        attr = analysis['winning_mean_attr']

        if feature_names is None:
            feature_names = [f'feature_{i}' for i in range(len(attr))]

        # Sort by absolute value
        sorted_idx = np.argsort(np.abs(attr))[::-1][:n_features]

        df = pd.DataFrame({
            'feature': [feature_names[i] for i in sorted_idx],
            'winning_attr': attr[sorted_idx],
        })

        if 'losing_mean_attr' in analysis:
            losing_attr = analysis['losing_mean_attr']
            df['losing_attr'] = losing_attr[sorted_idx]
            df['discriminative'] = df['winning_attr'] - df['losing_attr']

        return df


class AttributionFilteredBacktester(IGBacktester):
    """
    Backtester that filters trades based on attribution patterns.

    Rejects trades where attributions don't meet certain criteria,
    such as excessive dependence on volatile features.

    Parameters
    ----------
    model : TradingModelWithIG
        Trading model with IG capability
    initial_capital : float
        Starting capital
    feature_constraints : dict
        Constraints on feature attributions
        Example: {'sentiment': 0.3}  # max attribution for sentiment
    min_fundamental_support : float
        Minimum combined attribution from fundamental features
    fundamental_features : list of str
        Names of fundamental features
    """

    def __init__(
        self,
        model,
        initial_capital: float = 100_000,
        transaction_cost: float = 0.001,
        position_size: float = 0.1,
        feature_constraints: Optional[Dict[str, float]] = None,
        min_fundamental_support: float = 0.0,
        fundamental_features: Optional[List[str]] = None,
        feature_names: Optional[List[str]] = None,
    ):
        super().__init__(
            model=model,
            initial_capital=initial_capital,
            transaction_cost=transaction_cost,
            position_size=position_size,
            log_attributions=True,
        )

        self.feature_constraints = feature_constraints or {}
        self.min_fundamental_support = min_fundamental_support
        self.fundamental_features = fundamental_features or []
        self.feature_names = feature_names or []

        self.rejected_trades: List[Dict] = []

    def _should_execute_trade(
        self,
        attributions: np.ndarray,
        confidence: float,
    ) -> Tuple[bool, str]:
        """
        Check if trade should be executed based on attribution criteria.

        Returns
        -------
        should_execute : bool
            Whether to execute the trade
        reason : str
            Reason for decision
        """
        if attributions is None:
            return True, "No attributions available"

        # Check feature constraints
        for feature_name, max_attr in self.feature_constraints.items():
            if feature_name in self.feature_names:
                idx = self.feature_names.index(feature_name)
                if abs(attributions[idx]) > max_attr:
                    return False, f"Excessive dependence on {feature_name}: {attributions[idx]:.3f}"

        # Check fundamental support
        if self.fundamental_features and self.min_fundamental_support > 0:
            fundamental_attr = 0.0
            for feature_name in self.fundamental_features:
                if feature_name in self.feature_names:
                    idx = self.feature_names.index(feature_name)
                    fundamental_attr += abs(attributions[idx])

            if fundamental_attr < self.min_fundamental_support:
                return False, f"Insufficient fundamental support: {fundamental_attr:.3f}"

        return True, "Trade approved"

    def _generate_signal(
        self,
        features: np.ndarray,
        threshold: float = 0.5,
    ) -> Tuple[int, float, Optional[np.ndarray]]:
        """Generate signal with attribution filtering."""
        signal, confidence, attributions = super()._generate_signal(features, threshold)

        if signal != 0 and attributions is not None:
            should_execute, reason = self._should_execute_trade(attributions, confidence)

            if not should_execute:
                self.rejected_trades.append({
                    'original_signal': signal,
                    'confidence': confidence,
                    'attributions': attributions.copy(),
                    'reason': reason,
                })
                return 0, confidence, attributions  # Return neutral signal

        return signal, confidence, attributions


def calculate_trading_metrics(
    returns: Union[np.ndarray, pd.Series],
    risk_free_rate: float = 0.02,
) -> Dict[str, float]:
    """
    Calculate common trading performance metrics.

    Parameters
    ----------
    returns : array-like
        Return series
    risk_free_rate : float
        Annual risk-free rate

    Returns
    -------
    metrics : dict
        Performance metrics
    """
    if isinstance(returns, pd.Series):
        returns = returns.values

    returns = returns[~np.isnan(returns)]

    if len(returns) == 0:
        return {}

    # Daily risk-free rate
    rf_daily = risk_free_rate / 252

    # Excess returns
    excess_returns = returns - rf_daily

    metrics = {
        'total_return': np.prod(1 + returns) - 1,
        'annualized_return': np.mean(returns) * 252,
        'annualized_volatility': np.std(returns) * np.sqrt(252),
        'sharpe_ratio': np.mean(excess_returns) / np.std(returns) * np.sqrt(252) if np.std(returns) > 0 else 0,
        'max_drawdown': np.min(np.minimum.accumulate(np.cumprod(1 + returns)) / np.maximum.accumulate(np.cumprod(1 + returns)) - 1),
        'skewness': pd.Series(returns).skew(),
        'kurtosis': pd.Series(returns).kurtosis(),
    }

    # Sortino ratio
    downside = returns[returns < 0]
    if len(downside) > 0:
        downside_std = np.std(downside) * np.sqrt(252)
        metrics['sortino_ratio'] = (np.mean(returns) * 252 - risk_free_rate) / downside_std if downside_std > 0 else 0
    else:
        metrics['sortino_ratio'] = metrics['sharpe_ratio']

    # Calmar ratio
    if metrics['max_drawdown'] != 0:
        metrics['calmar_ratio'] = metrics['annualized_return'] / abs(metrics['max_drawdown'])
    else:
        metrics['calmar_ratio'] = 0

    return metrics

File: rust/src/backtest.rs
//! Backtesting Framework with Attribution Support
//!
//! This module provides backtesting capabilities for trading strategies
//! with integrated gradients attribution tracking.

use ndarray::Array1;
use serde::{Deserialize, Serialize};

use crate::integrated_gradients::IntegratedGradients;
use crate::model::NeuralNetwork;

/// Single trade record
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Trade {
    pub entry_index: usize,
    pub exit_index: usize,
    pub direction: i32,  // 1 for long, -1 for short
    pub entry_price: f64,
    pub exit_price: f64,
    pub pnl: f64,
    pub pnl_percent: f64,
    pub confidence: f64,
    pub attributions: Option<Vec<f64>>,
}

/// Backtest performance metrics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BacktestResults {
    pub total_return: f64,
    pub sharpe_ratio: f64,
    pub sortino_ratio: f64,
    pub max_drawdown: f64,
    pub win_rate: f64,
    pub profit_factor: f64,
    pub total_trades: usize,
    pub avg_trade_pnl: f64,
    pub equity_curve: Vec<f64>,
    pub trades: Vec<Trade>,
    pub winning_attr_mean: Option<Vec<f64>>,
    pub losing_attr_mean: Option<Vec<f64>>,
}

/// Backtester with integrated gradients support
pub struct Backtester<M: NeuralNetwork> {
    model: M,
    ig: IntegratedGradients,
    initial_capital: f64,
    transaction_cost: f64,
    position_size: f64,
    log_attributions: bool,
}

impl<M: NeuralNetwork> Backtester<M> {
    /// Create a new backtester
    ///
    /// # Arguments
    ///
    /// * `model` - Trading model
    /// * `initial_capital` - Starting capital
    /// * `transaction_cost` - Cost per trade as fraction
    /// * `position_size` - Position size as fraction of capital
    pub fn new(
        model: M,
        initial_capital: f64,
        transaction_cost: f64,
        position_size: f64,
    ) -> Self {
        Self {
            model,
            ig: IntegratedGradients::new(100),
            initial_capital,
            transaction_cost,
            position_size,
            log_attributions: true,
        }
    }

    /// Set whether to log attributions
    pub fn with_attributions(mut self, log: bool) -> Self {
        self.log_attributions = log;
        self
    }

    /// Set IG integration steps
    pub fn with_ig_steps(mut self, steps: usize) -> Self {
        self.ig = IntegratedGradients::new(steps);
        self
    }

    /// Generate trading signal from prediction
    fn generate_signal(&self, features: &Array1<f64>, threshold: f64) -> (i32, f64) {
        let output = self.model.forward(features);
        let prob = 1.0 / (1.0 + (-output[0]).exp());  // Sigmoid

        let signal = if prob > threshold {
            1  // Long
        } else if prob < (1.0 - threshold) {
            -1  // Short
        } else {
            0  // Neutral
        };

        (signal, prob)
    }

    /// Run backtest
    ///
    /// # Arguments
    ///
    /// * `features` - Feature matrix (n_samples x n_features)
    /// * `prices` - Price series
    /// * `signal_threshold` - Threshold for generating signals
    /// * `holding_period` - Number of bars to hold position
    ///
    /// # Returns
    ///
    /// Backtest results including performance metrics and trades
    pub fn run(
        &self,
        features: &ndarray::Array2<f64>,
        prices: &[f64],
        signal_threshold: f64,
        holding_period: usize,
    ) -> BacktestResults {
        let n = features.nrows();
        let mut trades: Vec<Trade> = Vec::new();
        let mut equity_curve = vec![self.initial_capital];
        let mut capital = self.initial_capital;

        let mut position: Option<(usize, i32, f64, f64, Option<Vec<f64>>)> = None;
        let mut bars_held = 0;

        for i in 0..(n - holding_period) {
            let features_i = features.row(i).to_owned();
            let price = prices[i];

            // Check if we need to close position
            if let Some((entry_idx, direction, entry_price, confidence, attrs)) = position.clone() {
                bars_held += 1;
                if bars_held >= holding_period {
                    // Close position
                    let pnl_percent = direction as f64 * (price / entry_price - 1.0) - self.transaction_cost;
                    let pnl = pnl_percent * self.position_size * capital;

                    capital += pnl;

                    trades.push(Trade {
                        entry_index: entry_idx,
                        exit_index: i,
                        direction,
                        entry_price,
                        exit_price: price,
                        pnl,
                        pnl_percent,
                        confidence,
                        attributions: attrs,
                    });

                    position = None;
                    bars_held = 0;
                }
            }

            // Generate signal if no position
            if position.is_none() {
                let (signal, confidence) = self.generate_signal(&features_i, signal_threshold);

                if signal != 0 {
                    // Compute attributions if logging
                    let attrs = if self.log_attributions {
                        Some(self.ig.explain(&self.model, &features_i, None).to_vec())
                    } else {
                        None
                    };

                    capital -= capital * self.position_size * self.transaction_cost;
                    position = Some((i, signal, price, confidence, attrs));
                }
            }

            equity_curve.push(capital);
        }

        // Close any remaining position
        if let Some((entry_idx, direction, entry_price, confidence, attrs)) = position {
            let exit_price = prices[n - 1];
            let pnl_percent = direction as f64 * (exit_price / entry_price - 1.0) - self.transaction_cost;
            let pnl = pnl_percent * self.position_size * capital;

            capital += pnl;

            trades.push(Trade {
                entry_index: entry_idx,
                exit_index: n - 1,
                direction,
                entry_price,
                exit_price,
                pnl,
                pnl_percent,
                confidence,
                attributions: attrs,
            });
        }

        equity_curve.push(capital);

        // Calculate metrics
        self.calculate_results(equity_curve, trades)
    }

    /// Calculate performance metrics
    fn calculate_results(&self, equity_curve: Vec<f64>, trades: Vec<Trade>) -> BacktestResults {
        let n = equity_curve.len();

        // Calculate returns
        let returns: Vec<f64> = equity_curve
            .windows(2)
            .map(|w| w[1] / w[0] - 1.0)
            .collect();

        // Total return
        let total_return = equity_curve.last().unwrap() / self.initial_capital - 1.0;

        // Sharpe ratio (annualized, assuming daily data)
        let mean_return: f64 = returns.iter().sum::<f64>() / returns.len() as f64;
        let variance: f64 = returns.iter().map(|r| (r - mean_return).powi(2)).sum::<f64>()
            / returns.len() as f64;
        let std_dev = variance.sqrt();
        let sharpe_ratio = if std_dev > 0.0 {
            mean_return / std_dev * (252.0_f64).sqrt()
        } else {
            0.0
        };

        // Sortino ratio
        let downside_returns: Vec<f64> = returns.iter().filter(|&&r| r < 0.0).cloned().collect();
        let downside_std = if !downside_returns.is_empty() {
            let var: f64 = downside_returns.iter().map(|r| r.powi(2)).sum::<f64>()
                / downside_returns.len() as f64;
            var.sqrt()
        } else {
            std_dev
        };
        let sortino_ratio = if downside_std > 0.0 {
            mean_return / downside_std * (252.0_f64).sqrt()
        } else {
            sharpe_ratio
        };

        // Maximum drawdown
        let mut peak = self.initial_capital;
        let mut max_drawdown = 0.0;
        for &equity in &equity_curve {
            if equity > peak {
                peak = equity;
            }
            let drawdown = (peak - equity) / peak;
            if drawdown > max_drawdown {
                max_drawdown = drawdown;
            }
        }

        // Trade statistics
        let winning_trades: Vec<&Trade> = trades.iter().filter(|t| t.pnl > 0.0).collect();
        let losing_trades: Vec<&Trade> = trades.iter().filter(|t| t.pnl <= 0.0).collect();

        let win_rate = if !trades.is_empty() {
            winning_trades.len() as f64 / trades.len() as f64
        } else {
            0.0
        };

        let total_profit: f64 = winning_trades.iter().map(|t| t.pnl).sum();
        let total_loss: f64 = losing_trades.iter().map(|t| t.pnl.abs()).sum();
        let profit_factor = if total_loss > 0.0 {
            total_profit / total_loss
        } else {
            f64::INFINITY
        };

        let avg_trade_pnl = if !trades.is_empty() {
            trades.iter().map(|t| t.pnl).sum::<f64>() / trades.len() as f64
        } else {
            0.0
        };

        // Attribution analysis
        let (winning_attr_mean, losing_attr_mean) = self.analyze_attributions(&trades);

        BacktestResults {
            total_return,
            sharpe_ratio,
            sortino_ratio,
            max_drawdown: -max_drawdown,
            win_rate,
            profit_factor,
            total_trades: trades.len(),
            avg_trade_pnl,
            equity_curve,
            trades,
            winning_attr_mean,
            losing_attr_mean,
        }
    }

    /// Analyze attributions for winning vs losing trades
    fn analyze_attributions(&self, trades: &[Trade]) -> (Option<Vec<f64>>, Option<Vec<f64>>) {
        let winning: Vec<&Vec<f64>> = trades
            .iter()
            .filter(|t| t.pnl > 0.0 && t.attributions.is_some())
            .map(|t| t.attributions.as_ref().unwrap())
            .collect();

        let losing: Vec<&Vec<f64>> = trades
            .iter()
            .filter(|t| t.pnl <= 0.0 && t.attributions.is_some())
            .map(|t| t.attributions.as_ref().unwrap())
            .collect();

        let winning_mean = if !winning.is_empty() {
            let n_features = winning[0].len();
            let mut mean = vec![0.0; n_features];
            for attrs in &winning {
                for (i, &a) in attrs.iter().enumerate() {
                    mean[i] += a;
                }
            }
            for m in &mut mean {
                *m /= winning.len() as f64;
            }
            Some(mean)
        } else {
            None
        };

        let losing_mean = if !losing.is_empty() {
            let n_features = losing[0].len();
            let mut mean = vec![0.0; n_features];
            for attrs in &losing {
                for (i, &a) in attrs.iter().enumerate() {
                    mean[i] += a;
                }
            }
            for m in &mut mean {
                *m /= losing.len() as f64;
            }
            Some(mean)
        } else {
            None
        };

        (winning_mean, losing_mean)
    }
}

/// Calculate common trading metrics from returns
pub fn calculate_metrics(returns: &[f64]) -> std::collections::HashMap<String, f64> {
    let mut metrics = std::collections::HashMap::new();

    if returns.is_empty() {
        return metrics;
    }

    let n = returns.len() as f64;

    // Total return
    let total_return: f64 = returns.iter().map(|r| 1.0 + r).product::<f64>() - 1.0;
    metrics.insert("total_return".to_string(), total_return);

    // Mean return
    let mean_return: f64 = returns.iter().sum::<f64>() / n;
    metrics.insert("annualized_return".to_string(), mean_return * 252.0);

    // Volatility
    let variance: f64 = returns.iter().map(|r| (r - mean_return).powi(2)).sum::<f64>() / n;
    let volatility = variance.sqrt() * (252.0_f64).sqrt();
    metrics.insert("annualized_volatility".to_string(), volatility);

    // Sharpe ratio
    let sharpe = if volatility > 0.0 {
        mean_return * 252.0 / volatility
    } else {
        0.0
    };
    metrics.insert("sharpe_ratio".to_string(), sharpe);

    // Maximum drawdown
    let mut equity = 1.0;
    let mut peak = 1.0;
    let mut max_dd = 0.0;
    for r in returns {
        equity *= 1.0 + r;
        if equity > peak {
            peak = equity;
        }
        let dd = (peak - equity) / peak;
        if dd > max_dd {
            max_dd = dd;
        }
    }
    metrics.insert("max_drawdown".to_string(), -max_dd);

    // Skewness
    let skewness: f64 = returns
        .iter()
        .map(|r| ((r - mean_return) / variance.sqrt()).powi(3))
        .sum::<f64>()
        / n;
    metrics.insert("skewness".to_string(), skewness);

    // Kurtosis
    let kurtosis: f64 = returns
        .iter()
        .map(|r| ((r - mean_return) / variance.sqrt()).powi(4))
        .sum::<f64>()
        / n
        - 3.0;
    metrics.insert("kurtosis".to_string(), kurtosis);

    metrics
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::TradingModel;
    use ndarray::Array2;

    #[test]
    fn test_backtester_basic() {
        let model = TradingModel::new(5, vec![10], 1);
        let backtester = Backtester::new(model, 100_000.0, 0.001, 0.1);

        // Create simple test data
        let features = Array2::from_shape_fn((100, 5), |_| 0.1);
        let prices: Vec<f64> = (0..100).map(|i| 100.0 + i as f64 * 0.1).collect();

        let results = backtester.run(&features, &prices, 0.55, 5);

        assert!(results.equity_curve.len() > 0);
    }

    #[test]
    fn test_metrics_calculation() {
        let returns = vec![0.01, -0.005, 0.02, -0.01, 0.015];
        let metrics = calculate_metrics(&returns);

        assert!(metrics.contains_key("sharpe_ratio"));
        assert!(metrics.contains_key("max_drawdown"));
    }
}
